add check that datadir exists

sbailey opened the pull request:
This PR implements #196 to log a warning message if the desimodel data directory isn't installed.

I found that I needed to do this within desimodel/io.py instead of desimodel/__init__.py since the top level module is imported at pip installation time, and importing things like `desimodel.io.datadir` led to knock-on effects like needing to import `yaml` which doesn't work unless running with `pip install --no-build-isolation ...`.  As a result, the warning message will get printed if `desimodel.io` is imported (the part that needs the data directory) but will not be printed from a bare `import desimodel`. I think that's a feature not a bug.

@weaverba137 please check if this matches what you were suggesting.

Example from a fresh environment without desimodel installed yet:
```
# confirm that desimodel isn't installed
(desitest) [check_datadir*] desimodel $ python -c "import desimodel"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    import desimodel
ModuleNotFoundError: No module named 'desimodel'

# install desimodel
(desitest) [check_datadir*] desimodel $ pip install .
Processing /Users/sbailey/desi/git/desimodel
...
Successfully installed desimodel-0.19.3.dev830

# importing bare desimodel should be ok
(desitest) [check_datadir*] desimodel $ python -c "import desimodel"

# but importing desimodel.io should print a warning
(desitest) [check_datadir*] desimodel $ python -c "import desimodel.io"
WARNING:io.py:787:_check_datadir: Missing /Users/sbailey/miniforge3/envs/desitest/lib/python3.13/site-packages/desimodel/data; please run "install_desimodel_data" from the commandline or "desimodel.install.main()" from a python prompt

# installing desimodel data makes the warning message go away
(desitest) [check_datadir*] desimodel $ install_desimodel_data
Installing desimodel data trunk to /Users/sbailey/miniforge3/envs/desitest/lib/python3.13/site-packages/desimodel
(desitest) [check_datadir*] desimodel $ python -c "import desimodel.io"

[no warning message printed]
```

sbailey committed:
add check that datadir exists

weaverba137 commented:
Quick reply: I think only raising the warning when `desimodel.io` is imported is fine. Have you also tested `from desimodel.io import function`?

I had in mind emitting `warnings.warn()`, which may be easier for automation to catch than just emitting a log warning.

weaverba137 reviewed (approved):
The code changes look fine, but still TBD: 

* Do we want to emit `log.warning()` or `warnings.warn()`?
* Does `from desimodel.io import function` also emit the warning?

sbailey commented:
> Quick reply: I think only raising the warning when `desimodel.io` is imported is fine. Have you also tested `from desimodel.io import function`?

Good idea to check that.  It also generates the warning:
```
(desitest) [check_datadir*] desimodel $ python -c "from desimodel.io import load_throughput"
WARNING:io.py:787:_check_datadir: Missing /Users/sbailey/miniforge3/envs/desitest/lib/python3.13/site-packages/desimodel/data; please run "install_desimodel_data" from the commandline or "desimodel.install.main()" from a python prompt
```

> I had in mind emitting `warnings.warn()`, which may be easier for automation to catch than just emitting a log warning.

I tried switching `log.warn` to `warnings.warn` and got a more obfuscated message:
```
(desitest) [check_datadir*] desimodel $ python -c "from desimodel.io import load_throughput"
/Users/sbailey/miniforge3/envs/desitest/lib/python3.13/site-packages/desimodel/io.py:787: UserWarning: Missing /Users/sbailey/miniforge3/envs/desitest/lib/python3.13/site-packages/desimodel/data; please run "install_desimodel_data" from the commandline or "desimodel.install.main()" from a python prompt
  warnings.warn(f'Missing {_datadir}; please run "install_desimodel_data" from the commandline or "desimodel.install.main()" from a python prompt')
```
i.e. it prints the UserWarning message as intended, but then it also prints the python line that generated the warning, which adds clutter and gets in the way of giving simple instructions to the user. Do you know if there is a way to use `warnings.warn` without it printing the second less informative line?  If the difference is only for testing purposes, I could update _check_datadir to return True/False if the directory is detected and trust that the warning is printed if False is returned.

weaverba137 commented:
OK, that's a fair point. Let's stick with `log.warning()` for now. But we should have some unit test for this function. You can patch `os.path.isdir()` to return True or False, and then see if the warning is emitted.

sbailey committed:
add desimodel github to warning; add tests

sbailey committed:
update warning

sbailey commented:
I added unit tests and added a reference to https://github.com/desihub/desimodel in the warning, but I think it is useful to include the core instruction in the warning itself instead of pointing the user to a separate webpage with pages of text just to find the one command to run. If we find that we're tripping on these instructions getting out of sync, we can change the warning to just point to the README as the sole source of information.

sbailey committed:
update changes.rst

sbailey committed:
update scipy intersphinx mapping
